test/clients.go

```diff
@@ -204,6 +208,57 @@ func (clients *ServingClients) Delete(routes, configs, services []string) []erro
 	return errs
 }
 
+// WaitForDeletion returns an error if any of the resources take too long to fully terminate.
+func (clients *ServingClients) WaitForDeletion(routes, configs, services []string, timeout time.Duration) error {
+	toWait := map[string]struct {
+		names []string
+		get   func(string) (interface{}, error)
+	}{
+		"routes": {
+			names: routes,
+			get: func(name string) (interface{}, error) {
+				return clients.Routes.Get(context.Background(), name, metav1.GetOptions{})
+			},
+		},
+		"configs": {
+			names: configs,
+			get: func(name string) (interface{}, error) {
+				return clients.Configs.Get(context.Background(), name, metav1.GetOptions{})
+			},
+		},
+		"services": {
+			names: services,
+			get: func(name string) (interface{}, error) {
+				return clients.Services.Get(context.Background(), name, metav1.GetOptions{})
+			},
+		},
+	}
+	return wait.PollImmediate(time.Second*2, timeout, func() (bool, error) {
+		total := len(routes) + len(services) + len(configs)
```

Comment:
fwiw rather than counting down from total I think this loop might be clearer/more concise if you just return false if you see any of the gets succeed (total untested pseudo-code, but seems like itd work:). Not a big deal by any means though, just a suggestion.

```
for resourceType, resource := range toWait {
  resourcesRemain := false
  for _, name := range resource.names {
    if name == "" { continue }
    
    _, err := resource.get(name); 
    if err != nil && !errors.IsNotFound(err) {
      return true, err
    }
    if err == nil {
      log.Printf("waiting on resource deletion for resource %s/%s", resourceType, name)
      resourcesRemain = true
    }
  }
  
  return resourcesRemain, nil
```